app/services/lms_sync_service.py

```diff
@@ -86,6 +99,191 @@ async def sync_assignments(self):
+            except:
+                print("Skipping over student not in LDAP: ", pid or email or "<unknown>")
+                return None
+
+            try:
+                user = await self.user_service.get_user_by_onyen(user_info.onyen)
+                return ResourcePair(db_resource=user, lms_resource=canvas_user, user_info=user_info)
+
+            except UserNotFoundException:
+                return ResourcePair(db_resource=None, lms_resource=canvas_user, user_info=user_info)
+        
+
+        # We begin by checking if there are any users that have been deleted from Canvas.
+        # This is 1 DB call per user so doesn't really need optimization.
+        for db_user in db_users:
+            pid = await self.canvas_service.get_pid_from_onyen(db_user.onyen)
+            if pid not in canvas_user_pids:
+                pairs.append(ResourcePair(db_resource=db_user, lms_resource=None))
+        
+        # Then we process Canvas users
+        pairs += await asyncio.gather(*[process_canvas_user(u) for u in canvas_users])
+
+        return [p for p in pairs if p is not None]
+    
+    async def _pair_db_canvas_assignments(self, db_assignments: list[AssignmentModel], canvas_assignments: list[dict]):
+        pairs = []
+
+        canvas_assignment_ids = [a["id"] for a in canvas_assignments]
+        for db_assignment in db_assignments:
+            if db_assignment.id not in canvas_assignment_ids:
+                pairs.append(ResourcePair(db_resource=db_assignment, lms_resource=None))
+
+        for canvas_assignment in canvas_assignments:
+            try:
+                db_assignment = await self.assignment_service.get_assignment_by_id(canvas_assignment["id"])
+                pairs.append(ResourcePair(db_resource=db_assignment, lms_resource=canvas_assignment))
+
+            except AssignmentNotFoundException as e:
+                pairs.append(ResourcePair(db_resource=None, lms_resource=canvas_assignment))
+
+        return pairs
+    
+    async def sync_student(
+        self,
+        resource_pair: ResourcePair
+    ):
+        db_student, canvas_student = resource_pair.db_resource, resource_pair.lms_resource
+
+        if db_student is None and canvas_student is not None:
+            # create
+            await self.student_service.create_student(
+                onyen=resource_pair.user_info.onyen,
+                name=canvas_student["name"],
+                email=canvas_student["email"]
+            )
+            await self.canvas_service.associate_pid_to_user(resource_pair.user_info.onyen, resource_pair.user_info.pid)
+
+        if db_student is not None and canvas_student is not None:
+            # update
+            pass
+
+        if db_student is not None and canvas_student is None:
+            # delete
+            await self.student_service.delete_user(db_student.onyen)
+            try: await self.canvas_service.unassociate_pid_from_user(db_student.onyen)
```

Comment:
do we have something in the clean up service to handle this? 
If not, we can move this logic over into a single method to handle both student/instructor clean up - at a later date not important rn

Reply:
Handle what?

Reply:
Note: pid association is cascade deleted, the try/catch on the unassociate call can be deleted now